feat: support a function that accepts a component definition, added some test cases

liby opened the pull request:
Solved the two points 1 and 3 mentioned in #14 
> 1. Function components with `observer` decorator receive parameters: throw error
> 3. React.FunctionComponent will be skipped execution

gndelia commented:
Thanks for your contribution 🚀 !   
I'll review this PR in the afternoon, later today. Point 2 would be the only one pending, right?

gndelia commented on transform.test.ts:
```diff
@@ -50,6 +50,22 @@ const testCases: TestCase[] = [
+      const MyComponent = (props: Props2) => {
+        return <span>{props.id}</span>
+      }`,
+  },
```
Can we add a scenario where using `FunctionComponent` without props? Something that shows how this

```js
const MyComponent: React.FunctionComponent = () => <span>
```

is converted properly, for the sake of having more scenarios tested

gndelia commented on transform.test.ts:
```diff
@@ -381,6 +397,40 @@ const testCases: TestCase[] = [
+      })`,
+  },
+  {
+    input: `
+      import React from 'react';
+      import { observer } from "mobx-react-lite";
+
+      type Props = { id: number };
+      const MyComponent: React.FC<Props> = observer((props) => {
+        return <span>{props.id}</span>
+      })`,
+    output: `
+      import React from 'react';
+      import { observer } from "mobx-react-lite";
+
+      type Props = { id: number };
+      const MyComponent = observer((props: Props) => {
+        return <span>{props.id}</span>
+      })`,
+  },
```
Can we also test this scenario in combination with `React.SFC` and `React.FunctionComponent` ?

gndelia reviewed (changes requested):
Thank you @liby  for your PR 🚀 I really appreciate the time and effort you have put, it is looking great! I left some minor changes to the tests to increase the confidence of the suite.

In addition to that, could you update the Readme adding some examples of the scenario your PR is fixing? Perhaps the `observer` scenario should be generalized to using a function that accepts a component definition.


Last, there was another scenario mentioned in the PR. Are you going to address it as part of this PR, or another? 

Looking forward to merging this PR 🥇

liby committed:
feat: support for function that accepts a component definition, added some test cases

liby commented:
> Point 2 would be the only one pending, right?

Yes, I don't have time to deal with this for now.

> I left some minor changes to the tests to increase the confidence of the suite.

Thank you for your advice, which I have modified according to your suggestions.

> In addition to that, could you update the Readme adding some examples of the scenario your PR is fixing?

How about you look again now?

> Last, there was another scenario mentioned in the PR. Are you going to address it as part of this PR, or another?

The main modification of PR is to support a function that accepts a component definition, all other modifications are additional. I think it's okay to keep or split out, depending on your opinion.

In addition, in the latest code, I fixed the problem of missing other parameters.

<details>
<summary>Before fix</summary>

```tsx
// before codemod runs
import { FC, forwardRef } from 'react';

type Props = { id: number };
const MyComponent: FC<Props> = forwardRef(({ id }, ref) => {
  return <span>{id}</span>
})

// after codemod runs
import { FC, forwardRef } from 'react';

type Props = { id: number };
const MyComponent = forwardRef((
  {
    id
  }: Props
  // `ref` is missing
) => {
  return <span>{id}</span>
})
```
</details>

<details>
<summary>After fix</summary>

```tsx
// before codemod runs
import { FC, forwardRef } from 'react';

type Props = { id: number };
const MyComponent: FC<Props> = forwardRef(({ id }, ref) => {
  return <span>{id}</span>
})

// after codemod runs
import { FC, forwardRef } from 'react';

type Props = { id: number };
const MyComponent = forwardRef((
  {
    id
  }: Props,
  ref
) => {
  return <span>{id}</span>
})
```
</details>

gndelia commented:
Great work! Merged 🚀 

@all-contributors please add @liby for docs and code
